prow/plugins/size/gen_files_group.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package size
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+)
+
+const genConfigFile = ".generated_files"
+
+// ghFileClient scopes to the only relevant functionality we require of a github client.
+type ghFileClient interface {
+	GetFile(org, repo, filepath, commit string) ([]byte, error)
+}
+
+// GenFilesGroup is a logical collection of files. Check for a file's
+// inclusion in the group using the Match method.
+type GenFilesGroup struct {
+	Paths, FileNames, PathPrefixes, FilePrefixes map[string]bool
+}
+
+// NewGenFilesGroup reads the .generated_files file in the root of the repository
+// and any referenced path files (from "path-from-repo" commands).
+func NewGenFilesGroup(gc ghFileClient, owner, repo, sha string) (g *GenFilesGroup, parseErrs []*ParseError, err error) {
+	g = &GenFilesGroup{
+		Paths:        make(map[string]bool),
+		FileNames:    make(map[string]bool),
+		PathPrefixes: make(map[string]bool),
+		FilePrefixes: make(map[string]bool),
+	}
+
+	var bs []byte
+	bs, err = gc.GetFile(owner, repo, genConfigFile, sha)
+	if err != nil {
+		err = &IOError{err: fmt.Errorf("GetFile error: %v", err)}
+		return
+	}
+
+	var repoFiles []string
+	repoFiles, parseErrs, err = g.load(bytes.NewBuffer(bs))
+	if err != nil {
+		return
+	}
+	for _, f := range repoFiles {
+		bs, err = gc.GetFile(owner, repo, f, sha)
+		if err != nil {
+			return
+		}
+		if err = g.loadPaths(bytes.NewBuffer(bs)); err != nil {
+			return
+		}
+	}
+	return
+}
+
+// Use load to read a generated files config file, and populate g with the commands.
+// "paths-from-repo" commands are aggregated into repoPaths. It is the caller's
+// responsiblity to fetch these and load them via g.loadPaths.
+func (g *GenFilesGroup) load(r io.Reader) (repoPaths []string, parseErrs []*ParseError, err *IOError) {
```

Comment:
Just return one `error`. Users of the library can do a type assertion if they care what kind of error it is.